observability/grafana/grafana.py

```diff
@@ -90,6 +91,84 @@ def fs_get_dashboard_folders(search_root: str) -> list[str]:
     ]
 
 
+def get_or_create_folder(
+    local_folder: str, g: GrafanaRunner, existing_folders: list[dict[str, any]]
+) -> str:
+    parts = local_folder.split("/")
+    if len(parts) != 2:
+        raise RuntimeError(
+            f"A 'grafana-dashboards' should be in the first subdirectory only. Error in {local_folder}"
+        )
+
+    folder_service_name = parts[0]
+
+    existing_uid = get_folder_uid(folder_service_name, existing_folders)
+    return (
+        existing_uid
+        if existing_uid != ""
+        else g.create_folder(folder_service_name).get("uid", "")
+    )
+
+
+def create_dashboard(
+    temp_file: str,
+    dashboard: dict[str, any],
+    folder_uid: str,
+    existing_dashboards: list[dict[str, any]],
+    g: GrafanaRunner,
+) -> None:
+    with open(temp_file, "w") as f:
+        dashboard["folderUid"] = folder_uid
+        json.dump(dashboard, f)
+
+    dashboard_found = [
+        e
+        for e in existing_dashboards
+        if e["title"] == dashboard["dashboard"]["title"]
+        if e["folderUid"] == folder_uid
+    ]
+    create_or_update = True
+    if dashboard_found:
+        assert len(dashboard_found) == 1
+        existing_dashboard = g.show_existing_dashboard(dashboard_found[0]["uid"])
+
+        # Deleting info that might change
+        # This script uses override and will always create new versions/ids
+        if existing_dashboard["dashboard"].get("uid", None):
+            del existing_dashboard["dashboard"]["uid"]
+        if existing_dashboard["dashboard"].get("id", None):
+            del existing_dashboard["dashboard"]["id"]
+        if existing_dashboard["dashboard"].get("version", None):
+            del existing_dashboard["dashboard"]["version"]
+        if dashboard["dashboard"].get("id", None):
+            del dashboard["dashboard"]["id"]
+        if dashboard["dashboard"].get("version", None):
+            del dashboard["dashboard"]["version"]
+
+        if existing_dashboard["dashboard"] == dashboard["dashboard"]:
+            print("Dashboard matches, no update needed")
+            create_or_update = False
+
+    if create_or_update:
+        print("Dashboard differs or does not exist update needed")
+        g.create_dashboard(temp_file)
+
+
+def delete_stale_dashboards(
+    d: str,
+    dashboards_visited: set[str],
+    existing_folders: list[dict[str, any]],
+    g: GrafanaRunner,
+) -> None:
+    k = f"{d['folderUid']}_{d['title']}"
+    if k not in dashboards_visited:
+        for amf in AZURE_MANAGED_FOLDERS:
+            uid = get_folder_uid(amf, existing_folders)
+            if uid and uid in k:
+                return
+        g.delete_dashboard(k.split("_")[1])
```

Comment:
* should this function be named `delete_stale_dashboard` (singular)?
* from what i understand you only need `k` to check if this dashboard still exists in the repo
* instead of `uid in k` you could directly check `uid == d['folderUid']`
* instead `k.split("_")[1]` you can directly use `d['title']`

Reply:
makes sense, fixed!